src/Diagnostics.DataProviders/LogDecoratorBase.cs

```diff
@@ -0,0 +1,85 @@
+﻿using System;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Diagnostics.Logger;
+using Diagnostics.ModelsAndUtils.Models;
+
+namespace Diagnostics.DataProviders
+{
+    internal abstract class LogDecoratorBase : IMetadataProvider
+    {
+        protected string _requestId;
+        private DataProviderMetadata _metadataProvider;
+        private CancellationToken _dataSourceCancellationToken;
+
+        protected LogDecoratorBase(DataProviderContext context, DataProviderMetadata metaData)
+        {
+            _metadataProvider = metaData;
+            _requestId = context.RequestId;
+            _dataSourceCancellationToken = context.DataSourcesCancellationToken;
+        }
+
+        protected async Task<T> MakeDependencyCall<T>(Task<T> dataProviderTask, [CallerMemberName]string dataProviderOperation = "")
+        {
+            Exception dataProviderException = null;
+            DateTime startTime = DateTime.UtcNow, endTime;
+            CancellationTokenRegistration cTokenRegistration;
+
+            try
+            {
+                var tcs = new TaskCompletionSource<bool>();
+                cTokenRegistration = _dataSourceCancellationToken.Register(() => tcs.TrySetResult(true));
+                var completedTask = await Task.WhenAny(new Task[] { dataProviderTask, tcs.Task });
+
+                if (completedTask.Id == dataProviderTask.Id)
+                {
+                    return await dataProviderTask;
+                }
+                else
+                {
+                    var dataSourceName = this.GetType().Name;
+                    var logDecoratorSuffix = "LogDecorator";
+
+                    if (dataSourceName.EndsWith(logDecoratorSuffix, StringComparison.InvariantCultureIgnoreCase))
+                    {
+                        dataSourceName = dataSourceName.Substring(0, dataSourceName.Length - logDecoratorSuffix.Length);
+                    }
+                    throw new TimeoutException($"DataSource timed out: {dataSourceName}");
```

Comment:
Can we say `DataSource timed out: {dataSourceName}. DataProviderOperation: {dataProviderOperation}`. Observer and Geomaster have multiple operations so this will boil things down to which operations might be timing out in those data providers and if we want to look at it from a higher level we can just parse our the DataSourceName in Kusto and summarize on top of that.